### DISCUSSION TITLE
Help in creating Adversarial Environment

### DISCUSSION THREAD

**Discussion by AizazSharif** [praveen-palanisamy/macad-gym#37] [Ideas]
Hi @praveen-palanisamy,

I wanted to try the adversarial multi-agent example as mentioned in the related paper, but there are only 2 available examples and an adversarial example is not in it.
Kindly help in how to create such env and use it for training and testing.

Any information would be helpful.
Thanks

**Comment by praveen-palanisamy** [comment]
Hi @AizazSharif ,

An adversarial scenario was not explicitly evaluated/studied in the paper but it is easy to add one to MACAD-Gym.

1. [Sample guide to create and new learning environment or scenario](https://github.com/praveen-palanisamy/macad-gym/wiki/creating-new-learning-environments.md). This gives you the list of changes to create your own custom MACAD-Gym environment with a custom scenario.
2. Specifically for the adversarial multi-agent environment you are interested in, you can re-use [this urban driving through a signalized intersection scenario](https://github.com/praveen-palanisamy/macad-gym/blob/d0a2a7544345a1cd9479c9f66fc3ae0f25cccfbd/src/macad_gym/envs/intersection/urban_signal_intersection_3c.py) where 3 cars are crossing the intersection. To make the Actor `car2` behave adversarially to the `car1` Actor for example, you could change the following reward function spec for `car2`:
https://github.com/praveen-palanisamy/macad-gym/blob/d0a2a7544345a1cd9479c9f66fc3ae0f25cccfbd/src/macad_gym/envs/intersection/urban_signal_intersection_3c.py#L57
to use your own custom reward function or simply create a negative version (mini-max's counter part) of `car1`'s reward function.

For your reference, the Reward functions are implemented in [reward.py](https://github.com/praveen-palanisamy/macad-gym/blob/38884ac4bc7fb2e91a865950cff4eeadeed02f60/src/macad_gym/carla/reward.py).

I will be happy to help you if you need any further assistance in putting this together as a pull request.

**Comment by AizazSharif** [comment]
Hi @praveen-palanisamy,
Thanks for the help. I will go through the instructions and will update here soon. I asked about the Adversarial example since it's mentioned in the naming convention by the name '_Advrs_'. 

In  [urban driving through a signalized intersection scenario ](https://github.com/praveen-palanisamy/macad-gym/blob/d0a2a7544345a1cd9479c9f66fc3ae0f25cccfbd/src/macad_gym/envs/intersection/urban_signal_intersection_3c.py) I wanted to ask if the multi-agents training are done through shared policy or separate. I am interested in working on separate policy training.

**Comment by praveen-palanisamy** [comment]
Yes, you could use separate policies for each of the agents.

An example where different policy parameters are used for each Car actor is available in the [MACAD-Agents](https://github.com/praveen-palanisamy/macad-agents) repository. It's implemented using Ray and specifically, the following lines shows how you could use different policy parameters for each of the agents:

https://github.com/praveen-palanisamy/macad-agents/blob/c7e48dd3dd1b9ce10dd5d52fc940300e3e328d67/src/macad_agents/rllib/impala_agent_independent.py#L312-L319

```python
                "multiagent": {
                    "policy_graphs": {
                        id: default_policy()
                        for id in env_actor_configs["actors"].keys()
                    },
                    "policy_mapping_fn":
                    tune.function(lambda agent_id: agent_id),
                },
```

You can also use a custom/different Deep RL algorithm for each or the agents to train the policy.

**Comment by AizazSharif** [comment]
This is great thanks a lot for the help @praveen-palanisamy. Hopefully, I am able to do this before I create a feature pull request.

**Comment by AizazSharif** [comment]
Hi @praveen-palanisamy ,
I was going through the framework and was wondering how to change Deep RL policies within Macad-gym. It's visible in macad-agents (Impala, PPO) but I was unable to find it in macad-gym environments to add my own. 

Any information would be helpful.
Thanks

**Comment by praveen-palanisamy** [comment]
Hey @AizazSharif ,
The Agent code which include the Deep RL policy implementation is kept out of the learning environment code [MACAD-Gym](https://github.com/praveen-palanisamy/macad-gym.git) to allow for modularity and plug-and-play with other RL libraries or RL environments. As you figured, sample Agent training scripts are provided in the [MACAD-Agents repository](https://github.com/praveen-palanisamy/macad-agents). 
MACAD-Agents works with MACAD-Gym and you can change the Deep RL Agent algorithms or policy definitions in [MACAD-Agents repository](https://github.com/praveen-palanisamy/macad-agents).
Hope that helps.

**Comment by AizazSharif** [comment]
Thanks for the reply @praveen-palanisamy. 
I am facing trouble while using Macad-agent impala examples. Even if I lower the resources the code runs but the cars don't move at all in default settings and without any training or crash Carla restarts the environment. Whereas in macad-gym, the agents make actions, and also we can have 2nd window which shows the agent's POV from the front. 

If you have any suggestions or pointer regarding this issue it will be really helpful.

Thanks.

**Comment by praveen-palanisamy** [comment]
Okay so that does sound to be specific to the the Impala Agent which has it's own resource requirements on top of the environment/CARLA. Since this is a different issue, can you open one  on the [MACAD-Agents](https://github.com/praveen-palanisamy/macad-agents) repository to keep things organized?

**Comment by praveen-palanisamy** [comment]
@AizazSharif : If you need further assistance on the original topic (Help in creating Adversarial Environments), we can continue it as an Idea Discussion.

**Comment by AizazSharif** [comment]
Hi @praveen-palanisamy ,
Thanks for moving it into a discussion. Regarding the Adversarial Environment, I wanted to ask how to add any gym (or stable-baseline) related model (PPO, etc) into basic_agent.py so that I can first train each agent in a multi-agent environment.

In macad-gym I can easily work on reward function, but without RL algorithm and training environment it looks difficult. 

Any help would be appreciated.
Thanks.

**Reply by praveen-palanisamy** [reply]
You can use stable-baselines or any RL Agent algorithm with MACAD-Gym. A good place to start simple is using [this example](https://github.com/praveen-palanisamy/macad-gym/blob/master/examples/basic_agent.py#L1-L41):
https://github.com/praveen-palanisamy/macad-gym/blob/38884ac4bc7fb2e91a865950cff4eeadeed02f60/examples/basic_agent.py#L1-L41

You can modify the `SimpleAgent` class or it's `get_action(...)` method to use any of the Agent class/methods from an RL library (preferably with built-in support for Multi-Agent RL environments).

**Reply by AizazSharif** [reply]
Yes I am currently using basic_agent.py to run my environments, but integrating the RL algorithm that supports multi-agent support seems an issue. Is there any work done that you can point out which can be integrated into macad gym (apart from macad agents)?

**Reply by AizazSharif** [reply]
This issue is highly dependent on #27 I think since you can import any TensorFlow-based model to create a training environment within macad-gym's **basic_agent.py** code.
